### PR TITLE
Fix: Problem with keras ResNet50 CAM visualisation #53

### PR DESCRIPTION
**Pull request by keisen**
**Cause**
In the `modify_model_backdrop` method, It is setting new Relu function to all the ‘activation’ attribute. i.e., It's all of Convolution Layer and Activation Layer.
In Keras's ResNet50, the `activation` attribute of each Conv layer is set to the Linear function and also has Activation layer.
Replacing this Linear function with the Relu function seems to be the cause of the gradient loss.
Actually, If there are not Activation Layer, this issue isn't occured.

**Detail**

Please refer to below comment ( https://github.com/raghakot/keras-vis/pull/122#issuecomment-406935791 ).

@raghakot ,
Since it is a bug fix, I didn’t add ~~tests and~~ documents.
Instead of that, I added an example.

https://github.com/keisen/keras-vis/blob/bugfix/%2353/examples/resnet/attention.ipynb

**( MODIFIED on 15 Aug 2018 )**

### CODE REVIEW

**Comment by keisen** [comment]
@raghakot ,
Sorry, I forgot to fix the test code.
I will fix it so please give me some time.

**Comment by keisen** [comment]
Hi, @raghakot 

Now, I'm fixing the test code.
I have a question regarding the following implementation.

https://github.com/raghakot/keras-vis/blob/e019cc43ce6c00b2151941b18dbad63164ad632a/vis/backend/tensorflow_backend.py#L87-L100

As a test, I removed this implementation.
But `GuidedBackprop` is working fine. (There is no problem with ReNet 50 @ Keras)

What is this implementation necessary for?
Is it important to use `tf.nn.relu` ?

Thanks.

**Comment by keisen** [comment]
I'm sorry. I misoperated.... ;-(

**Comment by keisen** [comment]
@raghakot, Please review this PR, and point out if there is a problem.
Also please tell me if you have a good idea about the test.

## Fixes

The problem can be avoided by modifying the following implementation.

https://github.com/raghakot/keras-vis/blob/e019cc43ce6c00b2151941b18dbad63164ad632a/vis/backend/tensorflow_backend.py#L87-L100

The problem is the condition of line 89.
You can avoid the problem by excluding `keras.activations.linear` from the condition.
(Change line 89 to `if hasattr(layer, 'activation’) and layer.activation != keras.activations.linear:` )

However, I deleted all of the above implementations in this PR.
If you will support Activation other than ReLU, you should add Activation with `gradient_override_map` instead of replacing Activation with ReLU.
(Ex: `gradient_override_map({‘Relu’: backdrop_modifier}, {‘Elu’: backdrop_modifier})` )
In this PR, addition of Activation is not implemented.


## Consultation about the test

I can not create a test to reproduce the problem because we have not identified the root cause of the problem.
The test case being committed is meaningless because it is too simple.

Since problem can be avoided by no replacing Activation, 
it seems to be a problem that the ReLU function is executed twice consecutively.
However, because it is `ReLU(x) = max(0, x)`, it should be `ReLU(ReLU(x)) = ReLU(x)`.
I don't think that this will have an adverse effect on the results.
Batch Normalization and Add merge layer  also do not seem to cause this problem.

Do you have any good idea about test ?


Regards

**Comment by keisen** [comment]
Hi,@raghakot.

I’ve thought about this PR, then I decided to merge it.
The reason is following.

- In the using backdrop_model with Resnet50, this issue is critical.
- This patch is able to resoluve build-error in travis-ci.

if it have any problem, please revert this PR.

p.s.
If you have any idea about the test of this PR, please response for https://github.com/raghakot/keras-vis/pull/122#issuecomment-406935791 .
